src/d3a/models/state.py

```diff
@@ -72,14 +73,50 @@ class ProsumptionInterface(StateInterface, ABC):
     def __init__(self):
         # Actual energy consumed/produced by the device at specific market slots
         self._energy_measurement_kWh: Dict[DateTime, float] = {}
+        self._unsettled_deviation_kWh: Dict[DateTime, float] = {}
+        self._forecast_measurement_deviation_kWh: Dict[DateTime, float] = {}
+
+    def _calculate_unsettled_energy_kWh(self, measured_energy_kWh, time_slot):
+        """Return the unsettled energy (produced or consumed) in kWh."""
+        return 0.0
 
     def set_energy_measurement_kWh(self, energy_kWh: float, time_slot: DateTime) -> None:
         """Set the actual energy consumed/produced by the device in the given market slot."""
         self._energy_measurement_kWh[time_slot] = energy_kWh
+        self._forecast_measurement_deviation_kWh[time_slot] = self._calculate_unsettled_energy_kWh(
+            energy_kWh, time_slot)
+        self._unsettled_deviation_kWh[time_slot] = \
+            abs(self._forecast_measurement_deviation_kWh[time_slot])
 
     def get_energy_measurement_kWh(self, time_slot: DateTime) -> float:
         """Get the actual energy consumed/produced by the device in the given market slot."""
-        return self._energy_measurement_kWh.get(time_slot, None)
+        return self._energy_measurement_kWh.get(time_slot)
+
+    def get_forecast_measurement_deviation_kWh(self, time_slot: DateTime) -> float:
+        """Get the energy deviation of forecasted energy from measurement by the device in
+        the given market slot. Negative value means that the deviation is beneficial to the
+        grid (and can be posted as an offer), positive value means that the deviation is
+        detrimental to the grid (and can be posted as a bid)"""
+        return self._forecast_measurement_deviation_kWh.get(time_slot)
+
+    def should_post_bid(self, time_slot: DateTime) -> bool:
+        return self._forecast_measurement_deviation_kWh.get(time_slot) > 0.0
+
+    def should_post_offer(self, time_slot: DateTime) -> bool:
+        return self._forecast_measurement_deviation_kWh.get(time_slot) < 0.0
+
+    def get_unsettled_deviation_kWh(self, time_slot: DateTime) -> float:
+        """Get the unsettled energy deviation of forecasted energy from measurement by the device
+        in the given market slot."""
+        return self._unsettled_deviation_kWh.get(time_slot)
+
+    def decrement_unsettled_deviation(
+            self, purchased_energy_kWh: float, time_slot: DateTime) -> None:
+        """Decrease the device unsettled energy in a specific market slot."""
+        self._unsettled_deviation_kWh[time_slot] -= purchased_energy_kWh
+        assert self._unsettled_deviation_kWh[time_slot] >= -FLOATING_POINT_TOLERANCE, (
```

Comment:
I suggest to make this potentially more flexible, in case we decide to start using +/- values for consumption/production (as we discussed some months ago):
```suggestion
        assert abs(self._unsettled_deviation_kWh[time_slot]) >= FLOATING_POINT_TOLERANCE, (
```

Reply:
This check is due to the floating point precision (or lack thereof) of Python. For instance, assuming that the value of the _unsettled_deviation_kWh[time_slot] == 0.25 , and purchased_energy_kWh == 0.25 . In some cases, due to floating point precision errors, this operation `self._unsettled_deviation_kWh[time_slot] -= purchased_energy_kWh` might lead to a minuscule negative value (e.g. 1e-18). In this case the >= 0 assertion would fail, even though all these operations were correct. Therefore we are obliged to include the minuscule negative values in our assertion, thus the - is really needed here